### PR TITLE
Fixes #141 - Garbage collector to delete performance data older than 14days

### PR DESCRIPTION
**Pull request by patilsuraj767**

### CODE REVIEW

File: ros/processor/garbage_collector.py [hunk 1]
```diff
@@ -0,0 +1,33 @@
+from ros.lib.app import app, db
+from ros.lib.models import PerformanceProfile
+from datetime import datetime, timedelta
+from ros.lib.config import GARBAGE_COLLECTION_INTERVAL
+import time
+import logging
+
+logging.basicConfig(
```
Comment (kgaikwad) [inline ros/processor/garbage_collector.py:8]: It seems rest of the processors also using the same configs. 
Will it be possible to have a separate helper function which does logger initialization?

File: ros/lib/config.py [hunk 2]
```diff
@@ -25,3 +25,5 @@
 APP_NAME = os.getenv("APP_NAME", "ros")
 
 INSIGHTS_EXTRACT_LOGLEVEL = os.getenv("INSIGHTS_EXTRACT_LOGLEVEL", "ERROR")
+
+GARBAGE_COLLECTION_INTERVAL = os.getenv("GARBAGE_COLLECTION_INTERVAL", 86400)
```
Comment (kgaikwad) [inline ros/lib/config.py:29]: Could you please add one sentence about this variable?

File: ros/processor/garbage_collector.py [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+
+class GarbageCollector():
+    def run(self):
+        while True:
+            self.remove_outdated_data()
+
+    def remove_outdated_data(self):
+        with app.app_context():
+            results = db.session.query(PerformanceProfile).filter(
+                      PerformanceProfile.report_date < datetime.today() - timedelta(days=10)).delete()
```
Comment (kgaikwad) [inline ros/processor/garbage_collector.py:23]: I think it would be good to make days value configurable. What do you think?

**Review by kgaikwad — CHANGES_REQUESTED** [review]
Thank you @patilsuraj767!
Added few inline comments.

File: ros/lib/config.py (hunk 2 above)
Comment (patilsuraj767) [reply ros/lib/config.py:31]: Added comment in the code.

File: ros/lib/config.py [hunk 4]
```diff
@@ -25,3 +26,16 @@
 APP_NAME = os.getenv("APP_NAME", "ros")
 
 INSIGHTS_EXTRACT_LOGLEVEL = os.getenv("INSIGHTS_EXTRACT_LOGLEVEL", "ERROR")
+
+# Time interval after which garbage collector is involved to check for outdated data.
+GARBAGE_COLLECTION_INTERVAL = os.getenv("GARBAGE_COLLECTION_INTERVAL", 86400)
+# Number of days after which data is considered to be outdated.
+DAYS_DATA_OUTDATED = os.getenv("DAYS_DATA_OUTDATED", 14)
```
Comment (kgaikwad) [inline ros/lib/config.py:33]: Will it be good idea to rename this variable to something like -
 `METRICS_PURGE_DAYS` or `DAYS_UNTIL_STALE` ?

File: ros/processor/garbage_collector.py [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+        while True:
+            self.remove_outdated_data()
+
+    def remove_outdated_data(self):
+        with app.app_context():
+            results = db.session.query(PerformanceProfile).filter(
+                      PerformanceProfile.report_date < datetime.today() - timedelta(days=DAYS_DATA_OUTDATED)).delete()
+            db.session.commit()
+            if results:
+                LOG.info("Deleted %s outdated performance profiles", results)
```
Comment (kgaikwad) [inline ros/processor/garbage_collector.py:21]:
```suggestion
                LOG.info("Deleted %s performance profiles older than %d days", results, DAYS_DATA_OUTDATED)
```

**Review by kgaikwad — CHANGES_REQUESTED** [review]

File: ros/lib/config.py (hunk 4 above)
Comment (apuntamb) [reply ros/lib/config.py:33]: Why 14, if already deleting data older than 10 days?

Comment (patilsuraj767) [reply ros/lib/config.py:33]: @apuntamb We have decided to delete the data older than 14 days to make it consistent with host inventory.

Comment (apuntamb) [reply ros/lib/config.py]: Sounds good, can we modify the title of the PR then?

File: ros/processor/garbage_collector.py [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+
+class GarbageCollector():
+    def run(self):
+        while True:
+            self.remove_outdated_data()
+
+    def remove_outdated_data(self):
+        with app.app_context():
+            results = db.session.query(PerformanceProfile).filter(
+                      PerformanceProfile.report_date < datetime.today() - timedelta(days=DAYS_UNTIL_STALE)).delete()
```
Comment (kgaikwad) [inline ros/processor/garbage_collector.py:18]:
```suggestion
                      PerformanceProfile.report_date < datetime.now(timezone.utc) - timedelta(days=DAYS_UNTIL_STALE)).delete()
```

@patilsuraj767 - As `report_date` stored in UTC, do we need to consider `datetime.now(timezone.utc)1 over `datetime.today()`?

**Comment by apuntamb** [comment]
@kgaikwad mind opening a new PR to remove conflicts and closing this one in favor of the former?

File: ros/lib/config.py (hunk 4 above)
Comment (kgaikwad) [reply ros/lib/config.py]: @apuntamb, any thoughts on renaming this variable or should I keep as it is?

File: ros/processor/garbage_collector.py [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+                                         datetime.now(timezone.utc) - timedelta(days=DAYS_UNTIL_STALE))).delete()
+            db.session.commit()
+            if results:
+                LOG.info("Deleted %s performance profiles older than %d days", results, DAYS_UNTIL_STALE)
+
+            time.sleep(GARBAGE_COLLECTION_INTERVAL)
+
+
+if __name__ == "__main__":
+    collector = GarbageCollector()
```
Comment (kgaikwad) [inline ros/processor/garbage_collector.py:28]: As already moved logic to file `ros/processor/main.py`, still require to run from here?

File: ros/lib/config.py (hunk 4 above)
Comment (apuntamb) [reply ros/lib/config.py]: How about `DAYS_SINCE_OUTDATED` or `DAYS_SINCE_STALE` ?

File: ros/processor/garbage_collector.py (hunk 7 above)
Comment (apuntamb) [reply ros/processor/garbage_collector.py:28]: Looks like a leftover.

**Comment by kgaikwad** [comment]
Closing this PR on behalf of new PR #84

**Pull request closed without merging**